packages/cli/src/sso.ee/saml/saml.service.ee.ts

```diff
@@ -236,20 +241,58 @@ export class SamlService {
 		};
 	}
 
-	async setSamlPreferences(prefs: Partial<SamlPreferences>): Promise<SamlPreferences | undefined> {
+	async setSamlPreferences(
+		prefs: Partial<SamlPreferences>,
+		tryFallback: boolean = false,
+	): Promise<SamlPreferences | undefined> {
 		await this.loadSamlify();
 		await this.loadPreferencesWithoutValidation(prefs);
 		if (prefs.metadataUrl) {
-			const fetchedMetadata = await this.fetchMetadataFromUrl();
-			if (fetchedMetadata) {
-				this._samlPreferences.metadata = fetchedMetadata;
+			try {
+				const fetchedMetadata = await this.fetchMetadataFromUrl();
+				if (fetchedMetadata) {
+					this._samlPreferences.metadata = fetchedMetadata;
+				} else {
+					// in this case the metadata url didn't produce a valid metadata for SAML
+					// therefore we are rejecting the change to it
+					throw new InvalidSamlMetadataUrlError();
+				}
+			} catch (error) {
+				if (tryFallback) {
+					// we were not able to produce correct metadata from the URL, but
+					// in this case we don't care and try to fallback on the saved metadata in the
+					// database.
+					this.logger.warn(
+						'SAML initialization detected an invalid metadata URL in database. Trying to initialize from metadata in database if available.',
+					);
+				} else {
+					throw error;
+				}
 			}
 		} else if (prefs.metadata) {
 			const validationResult = await this.validator.validateMetadata(prefs.metadata);
 			if (!validationResult) {
 				throw new InvalidSamlMetadataError();
 			}
 		}
+		// If SAML login is enabled, we need to ensure that we have valid metadata available
+		// if the metadata url is provided and it was possible to fetch and validate that metadata
+		// it is now stored in this._samlPreferences.metadata.
+		// if no metadata url was provided but metadata directly as XML, it is also already stored
+		// in this._samlPreferences.metadata.
+		if (isSamlLoginEnabled()) {
+			if (this._samlPreferences.metadata) {
+				const validationResult = await this.validator.validateMetadata(
+					this._samlPreferences.metadata,
+				);
+				if (!validationResult) {
+					throw new InvalidSamlMetadataError();
+				}
+			} else {
+				// in this case SAML login is enabled but no valid metadata is available
+				throw new InvalidSamlMetadataError();
```

Comment:
This is not covered by a test as far as I can tell.

![image](https://github.com/user-attachments/assets/7e317e3c-0cf8-4594-a359-2976fffe491a)

Reply:
I think this case currently can't happen, because `this._samlPreferences.metadata` is initialized with an empty string and can't be set to undefined. I would suggest to still keep this in, because in case we change the default value of `this._samlPreferences` at some point, so that the metadata property can become undefined, this case could happen.

Reply:
Yeah, unfortunately `''` is treated as falsy in JS. 

I added a test for it though 😸 

Refactoring the types and the structure would be the way to go here in the long run of course.